File: languages/haskell/flipper/src/Flipper/Internal/USART.hs
{-|
Module      : Flipper.Internal.USART
Description : Internal USART Module
Copyright   : George Morgan, Travis Whitaker 2016
License     : All rights reserved.
Maintainer  : travis@flipper.io
Stability   : Provisional
Portability : Windows, POSIX

-}

module Flipper.Internal.USART (
    configure
  , enable
  , disable
  , ready
  , put
  , get
  , push
  , pull
  ) where

import Data.Word

import Flipper.Internal.Buffer
import Flipper.Internal.Utils

import Foreign.ForeignPtr
import Foreign.Marshal.Utils
import Foreign.Ptr

configure :: IO Bool
configure = retSuc <$> c_usart_configure

enable :: IO ()
enable = c_usart_enable

disable :: IO ()
disable = c_usart_disable

ready :: IO Bool
ready = toBool <$> c_usart_ready

put :: Word8 -> IO ()
put = c_usart_put

get :: IO Word8
get = c_usart_get

push :: Buffer -> IO ()
push (Buffer p o l) = withForeignPtr p $ \p' ->
   c_usart_push (plusPtr p' o) (fromIntegral l)

pull :: Int -> IO Buffer
pull l
    | l <= 0    = error "pull: length must be greater than zero."
    | otherwise = do b@(Buffer p _ _) <- allocBufferSafe l
                     withForeignPtr p (\p' -> c_usart_pull p' (fromIntegral l))
                     return b

foreign import ccall safe "flipper/usart/usart.h usart_configure"
    c_usart_configure :: IO Word32

foreign import ccall safe "flipper/carbon/modules/usart.h usart_enable"
    c_usart_enable :: IO ()

foreign import ccall safe "flipper/carbon/modules/usart.h usart_disable"
    c_usart_disable :: IO ()

foreign import ccall safe "flipper/carbon/modules/usart.h usart_ready"
    c_usart_ready :: IO Word8

foreign import ccall safe "flipper/carbon/modules/usart.h usart_put"
    c_usart_put :: Word8 -> IO ()

foreign import ccall safe "flipper/carbon/modules/usart.h usart_get"
    c_usart_get :: IO Word8

foreign import ccall safe "flipper/carbon/modules/usart.h usart_push"
    c_usart_push :: Ptr Word8 -> Word32 -> IO ()

foreign import ccall safe "flipper/carbon/modules/usart.h usart_pull"
    c_usart_pull :: Ptr Word8 -> Word32 -> IO ()

File: boards/carbon/hal/modules/usart.c
#define __private_include__
#include <flipper/libflipper.h>
#include <flipper/carbon/usart.h>

int usart_configure(void) {
	lf_invoke(&_usart, _usart_configure, NULL);
	return lf_success;
}

void usart_enable(void) {
	lf_invoke(&_usart, _usart_enable, NULL);
}

void usart_disable(void) {
	lf_invoke(&_usart, _usart_disable, NULL);
}

uint8_t usart_ready(void) {
	return lf_invoke(&_usart, _usart_ready, NULL) >> 16;
}

void usart_put(uint8_t byte) {
	lf_invoke(&_usart, _usart_put, fmr_args(fmr_infer(byte)));
}

uint8_t usart_get(void) {
	return lf_invoke(&_usart, _usart_get, NULL) >> 16;
}

int usart_push(void *source, lf_size_t length) {
	return lf_push(&_usart, _usart_push, source, length, NULL);
}

int usart_pull(void *destination, lf_size_t length) {
	return lf_pull(&_usart, _usart_pull, destination, length, NULL);
}
